### PR TITLE
Application Level Mutual TLS Support Feature

### PR DESCRIPTION
**Pull request by Abirame-ship-it**
Includes the implementation of the REST API , JMS Flows and Gateway Truststore Certificate Update

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/AlertSubscriptionsApiServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package org.wso2.carbon.apimgt.rest.api.store.v1.impl;
```
Comment (SanojPunchihewa) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/AlertSubscriptionsApiServiceImpl.java:1]: Shall we add the missing license headers? Please check other files as well

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIConsumer.java [hunk 2]
```diff
@@ -788,4 +779,70 @@ Set<SubscribedAPI> getPaginatedSubscribedAPIsByApplication(Application applicati
...
+    /**
+     * Method to check whether a certificate for the given alias is present in the trust store and the database.
+     *
+     * @param alias : The alias of the certificate.
+     * @return : True if a certificate is present, false otherwise.
+     * @throws APIManagementException :
+     */
+    boolean isCertificatePresent(int tenantId, String alias) throws APIManagementException;
+
+}
```
Comment (sajinieKavindya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIConsumer.java:848]: a new line is required

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java [hunk 3]
```diff
@@ -27,10 +27,7 @@
 import org.apache.commons.logging.LogFactory;
 import org.wso2.carbon.apimgt.api.APIManagementException;
 import org.wso2.carbon.apimgt.api.model.APIStatus;
-import org.wso2.carbon.apimgt.gateway.APILoggerManager;
-import org.wso2.carbon.apimgt.gateway.EndpointCertificateDeployer;
-import org.wso2.carbon.apimgt.gateway.GoogleAnalyticsConfigDeployer;
-import org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer;
+import org.wso2.carbon.apimgt.gateway.*;
```
Comment (sajinieKavindya) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:30]: Shall we revert this change to remove the wild card import? Let's change the applicable other places as well.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/ApplicationCertificateDeployer.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/ApplicationCertificateDeployer.java:2]: 2022

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/ApplicationCertificateDeployer.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            } finally {
+                if (tenantFlowStarted) {
+                    PrivilegedCarbonContext.endTenantFlow();
+                }
+            }
+
+        }
+    }
+
+    private CloseableHttpResponse invokeService(String endpoint, String tenantDomain)
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/ApplicationCertificateDeployer.java:116]: add some debug logs for all these methods. not only this method.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java [hunk 6]
```diff
@@ -285,6 +271,31 @@ private void handleNotificationMessage(String eventType, long timestamp, String
...
+        } else if (EventType.APPLICATION_CERTIFICATE_ADD.toString().equals(eventType) ||
+                EventType.APPLICATION_CERTIFICATE_REMOVE.toString().equals(eventType)) {
+            ApplicationCertificateEvent applicationCertificateEvent = new Gson().fromJson(eventJson,
+                    ApplicationCertificateEvent.class);
+            if (EventType.APPLICATION_CERTIFICATE_ADD.toString().equals(eventType)) {
+                try {
+                    new ApplicationCertificateDeployer(applicationCertificateEvent.getTenantDomain())
+                            .deployCertificate(applicationCertificateEvent.getUUID());
+                } catch (APIManagementException e) {
+                    log.error(e);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:283]: Add a proper error message and log

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 7]
```diff
@@ -18,13 +18,9 @@
...
 
-import java.io.File;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
 import javax.xml.namespace.QName;
+import java.io.File;
+import java.util.*;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java:23]: why remove the imports and add wildcard import?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java [hunk 8]
```diff
@@ -2563,44 +2536,12 @@ public static class GlobalCacheInvalidation {
...
-        REMOVE_APPLICATION_KEYMAPPING,
-        SCOPE_CREATE,
-        SCOPE_UPDATE,
-        SCOPE_DELETE,
-        ENDPOINT_CERTIFICATE_ADD,
-        ENDPOINT_CERTIFICATE_UPDATE,
-        ENDPOINT_CERTIFICATE_REMOVE,
-        GA_CONFIG_UPDATE,
-        UDATE_API_LOG_LEVEL
+        API_CREATE, API_UPDATE, API_DELETE, API_LIFECYCLE_CHANGE, APPLICATION_CREATE, APPLICATION_UPDATE, APPLICATION_DELETE, APPLICATION_REGISTRATION_CREATE, POLICY_CREATE, POLICY_UPDATE, POLICY_DELETE, SUBSCRIPTIONS_CREATE, SUBSCRIPTIONS_UPDATE, SUBSCRIPTIONS_DELETE, DEPLOY_API_IN_GATEWAY, REMOVE_API_FROM_GATEWAY, REMOVE_APPLICATION_KEYMAPPING, SCOPE_CREATE, SCOPE_UPDATE, SCOPE_DELETE, ENDPOINT_CERTIFICATE_ADD, ENDPOINT_CERTIFICATE_UPDATE, ENDPOINT_CERTIFICATE_REMOVE, APPLICATION_CERTIFICATE_ADD, APPLICATION_CERTIFICATE_REMOVE, GA_CONFIG_UPDATE, UDATE_API_LOG_LEVEL
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConstants.java:2544]: this line is too long. any reason for formatting this whole class? can you only add your changes plz

**Comment by chamilaadhi** [comment]
@[Abirame-ship-it](https://github.com/Abirame-ship-it) can you only format the sections you added and leave the rest of the code as it is. seems you have formatted all classes and it is hard to find your changes

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/ApplicationCertificateEvent.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    public String getUUID() {
+
+        return UUID;
+    }
+
+    public void setUUID(String UUID) {
+
+        this.UUID = UUID;
+    }
+//    public String getName() {
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/events/ApplicationCertificateEvent.java:63]: Let's remove the commented code

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ClientCertificatesApiServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package org.wso2.carbon.apimgt.internal.service.impl;
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/ClientCertificatesApiServiceImpl.java:1]: Missing license header

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CertificateMappingUtil.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
+package org.wso2.carbon.apimgt.rest.api.store.v1.mappings;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CertificateMappingUtil.java:1]: license header
